reframe/utility/__init__.py

```diff
@@ -164,24 +170,55 @@ def ppretty(value, htchar=' ', lfchar='\n', indent=4, basic_offset=0):
             return '{}'
 
         items = [
-            nlch + repr(key) + ': ' +
-            ppretty(value[key], htchar, lfchar, indent, basic_offset + 1)
-            for key in value
+            nlch + repr(key) + ': ' + ppretty2(value[key]) for key in value
         ]
         return '{%s}' % (','.join(items) + lfchar +
                          htchar * indent * basic_offset)
     elif isinstance(value, set):
         if value == set():
             return 'set()'
 
-        items = [
-            nlch + ppretty(item, htchar, lfchar, indent, basic_offset + 1)
-            for item in value
-        ]
+        items = [nlch + ppretty2(item) for item in value]
         return '{%s}' % (','.join(items) + lfchar +
                          htchar * indent * basic_offset)
     else:
-        return repr(value)
+        try:
+            return repr(value, basic_offset=basic_offset)
+        except TypeError:
+            # Not our custom repr()
+            return repr(value)
+
+
+def _tracked_repr(func):
+    objects = set()
+
+    @functools.wraps(func)
+    def _repr(obj, **kwargs):
+        addr = id(obj)
+        if addr in objects:
+            return f'{type(obj).__name__}(...)@{hex(addr)}'
+
+        # Do not track builtin objects
+        if hasattr(obj, '__dict__'):
+            objects.add(addr)
+
+        return func(obj, **kwargs)
+
+    return _repr
+
+
+@_tracked_repr
+def repr(obj, *, htchar=' ', lfchar='\n', indent=4, basic_offset=0):
```

Comment:
Actually, here I don't need all the other options... I could update the way I'm trying to call `repr()` in `ppretty()`.